olm/operator-controller/v1alpha1/clusterextension_types.go

```diff
@@ -0,0 +1,518 @@
+	//   - -some-namespace
+	//   - some-namespace-
+	//   - thisisareallylongnamespacenamethatisgreaterthanthemaximumlength
+	//   - some.namespace
+	//
+	// [RFC 1123]: https://tools.ietf.org/html/rfc1123
+	//
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	//+kubebuilder:validation:MaxLength:=63
+	//+kubebuilder:validation:XValidation:rule="self == oldSelf",message="namespace is immutable"
+	Namespace string `json:"namespace"`
+
+	// serviceAccount is a required reference to a ServiceAccount that exists
+	// in the installNamespace. The provided ServiceAccount is used to install and
+	// manage the content for the package specified in the packageName field.
+	//
+	// In order to successfully install and manage the content for the package,
+	// the ServiceAccount provided via this field should be configured with the
+	// appropriate permissions to perform the necessary operations on all the
+	// resources that are included in the bundle of content being applied.
+	ServiceAccount ServiceAccountReference `json:"serviceAccount"`
+
+	// preflight is an optional field that can be used to configure the preflight checks run before installation or upgrade of the content for the package specified in the packageName field.
+	//
+	// When specified, it overrides the default configuration of the preflight checks that are required to execute successfully during an install/upgrade operation.
+	//
+	// When not specified, the default configuration for each preflight check will be used.
+	//
+	//+optional
+	Preflight *PreflightConfig `json:"preflight,omitempty"`
+}
+
+// CatalogSource defines the required fields for catalog source.
+type CatalogSource struct {
+	// packageName is a reference to the name of the package to be installed
+	// and is used to filter the content from catalogs.
+	//
+	// This field is required, immutable and follows the DNS subdomain name
+	// standard as defined in [RFC 1123]. This means that valid entries:
+	//   - Contain no more than 253 characters
+	//   - Contain only lowercase alphanumeric characters, '-', or '.'
+	//   - Start with an alphanumeric character
+	//   - End with an alphanumeric character
+	//
+	// Some examples of valid values are:
+	//   - some-package
+	//   - 123-package
+	//   - 1-package-2
+	//   - somepackage
+	//
+	// Some examples of invalid values are:
+	//   - -some-package
+	//   - some-package-
+	//   - thisisareallylongpackagenamethatisgreaterthanthemaximumlength
+	//   - some.package
+	//
+	// [RFC 1123]: https://tools.ietf.org/html/rfc1123
+	//
+	//+kubebuilder:validation:MaxLength:=253
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
+	//+kubebuilder:validation:XValidation:rule="self == oldSelf",message="packageName is immutable"
+	PackageName string `json:"packageName"`
+
+	// version is an optional semver constraint (a specific version or range of versions). When unspecified, the latest version available will be installed.
```

Comment:
Flow is a little janky in the start/end transitions, but authentically represents the block contents well.  Here's an interim version (i.e. I'm still making updates): https://gist.github.com/grokspawn/e64ec44219464145b91e174e1483ffc8

Reply:
This looks pretty good I think:
```
k explain clusterextension.spec.source.catalog

GROUP:      olm.operatorframework.io
KIND:       ClusterExtension
VERSION:    v1alpha1

FIELD: catalog <Object>


DESCRIPTION:
    catalog is used to configure how information is sourced from a catalog.
    This field is required when sourceType is "Catalog", and forbidden
    otherwise.

FIELDS:
  channels	<[]string>
    channels is an optional reference to a set of channels belonging to
    the package specified in the packageName field.

    A "channel" is a package-author-defined stream of updates for an extension.

    Each channel in the list must follow the DNS subdomain standard
    as defined in [RFC 1123]. It must contain only lowercase alphanumeric
    characters,
    hyphens (-) or periods (.), start and end with an alphanumeric character,
    and be no longer than 253 characters.

    When specified, it is used to constrain the set of installable bundles and
    the automated upgrade path. This constraint is an AND operation with the
    version field. For example:
      - Given channel is set to "foo"
      - Given version is set to ">=1.0.0, <1.5.0"
      - Only bundles that exist in channel "foo" AND satisfy the version range
    comparison will be considered installable
      - Automatic upgrades will be constrained to upgrade edges defined by the
    selected channel

    When unspecified, upgrade edges across all channels will be used to identify
    valid automatic upgrade paths.

    Some examples of valid values are:
      - 1.1.x
      - alpha
      - stable
      - stable-v1
      - v1-stable
      - dev-preview
      - preview
      - community

    Some examples of invalid values are:
      - -some-channel
      - some-channel-
      - thisisareallylongchannelnamethatisgreaterthanthemaximumlength
      - original_40
      - --default-channel

    [RFC 1123]: https://tools.ietf.org/html/rfc1123

  packageName	<string> -required-
    packageName is a reference to the name of the package to be installed
    and is used to filter the content from catalogs.

    packageName is required, immutable, and follows the DNS subdomain standard
    as defined in [RFC 1123]. It must contain only lowercase alphanumeric
    characters,
    hyphens (-) or periods (.), start and end with an alphanumeric character,
    and be no longer than 253 characters.

    Some examples of valid values are:
      - some-package
      - 123-package
      - 1-package-2
      - somepackage

    Some examples of invalid values are:
      - -some-package
      - some-package-
      - thisisareallylongpackagenamethatisgreaterthanthemaximumlength
      - some.package

    [RFC 1123]: https://tools.ietf.org/html/rfc1123

  selector	<Object>
    selector is an optional field that can be used
    to filter the set of ClusterCatalogs used in the bundle
    selection process.

    When unspecified, all ClusterCatalogs will be used in
    the bundle selection process.

  upgradeConstraintPolicy	<string>
  enum: CatalogProvided, SelfCertified
    upgradeConstraintPolicy is an optional field that controls whether
    the upgrade path(s) defined in the catalog are enforced for the package
    referenced in the packageName field.

    Allowed values are: "CatalogProvided" or "SelfCertified", or omitted.

    When this field is set to "CatalogProvided", automatic upgrades will only
    occur
    when upgrade constraints specified by the package author are met.

    When this field is set to "SelfCertified", the upgrade constraints specified
    by
    the package author are ignored. This allows for upgrades and downgrades to
    any version of the package. This is considered a dangerous operation as it
    can lead to unknown and potentially disastrous outcomes, such as data
    loss. It is assumed that users have independently verified changes when
    using this option.

    When this field is omitted, the default value is "CatalogProvided".

  version	<string>
    version is an optional semver constraint (a specific version or range of
    versions). When unspecified, the latest version available will be installed.

    Acceptable version ranges are no longer than 64 characters.
    Version ranges are composed of comma- or space-delimited values and one or
    more comparison operators, known as comparison strings. Additional
    comparison strings can be added using the OR operator (||).

    # Range Comparisons

    To specify a version range, you can use a comparison string like ">=3.0,
    <3.6". When specifying a range, automatic updates will occur within that
    range. The example comparison string means "install any version greater than
    or equal to 3.0.0 but less than 3.6.0.". It also states intent that if any
    upgrades are available within the version range after initial installation,
    those upgrades should be automatically performed.

    # Pinned Versions

    To specify an exact version to install you can use a version range that
    "pins" to a specific version. When pinning to a specific version, no
    automatic updates will occur. An example of a pinned version range is
    "0.6.0", which means "only install version 0.6.0 and never
    upgrade from this version".

    # Basic Comparison Operators

    The basic comparison operators and their meanings are:
      - "=", equal (not aliased to an operator)
      - "!=", not equal
      - "<", less than
      - ">", greater than
      - ">=", greater than OR equal to
      - "<=", less than OR equal to

    # Wildcard Comparisons

    You can use the "x", "X", and "*" characters as wildcard characters in all
    comparison operations. Some examples of using the wildcard characters:
      - "1.2.x", "1.2.X", and "1.2.*" is equivalent to ">=1.2.0, < 1.3.0"
      - ">= 1.2.x", ">= 1.2.X", and ">= 1.2.*" is equivalent to ">= 1.2.0"
      - "<= 2.x", "<= 2.X", and "<= 2.*" is equivalent to "< 3"
      - "x", "X", and "*" is equivalent to ">= 0.0.0"

    # Patch Release Comparisons

    When you want to specify a minor version up to the next major version you
    can use the "~" character to perform patch comparisons. Some examples:
      - "~1.2.3" is equivalent to ">=1.2.3, <1.3.0"
      - "~1" and "~1.x" is equivalent to ">=1, <2"
      - "~2.3" is equivalent to ">=2.3, <2.4"
      - "~1.2.x" is equivalent to ">=1.2.0, <1.3.0"

    # Major Release Comparisons

    You can use the "^" character to make major release comparisons after a
    stable 1.0.0 version is published. If there is no stable version published,
    // minor versions define the stability level. Some examples:
      - "^1.2.3" is equivalent to ">=1.2.3, <2.0.0"
      - "^1.2.x" is equivalent to ">=1.2.0, <2.0.0"
      - "^2.3" is equivalent to ">=2.3, <3"
      - "^2.x" is equivalent to ">=2.0.0, <3"
      - "^0.2.3" is equivalent to ">=0.2.3, <0.3.0"
      - "^0.2" is equivalent to ">=0.2.0, <0.3.0"
      - "^0.0.3" is equvalent to ">=0.0.3, <0.0.4"
      - "^0.0" is equivalent to ">=0.0.0, <0.1.0"
      - "^0" is equivalent to ">=0.0.0, <1.0.0"

    # OR Comparisons
    You can use the "||" character to represent an OR operation in the version
    range. Some examples:
      - ">=1.2.3, <2.0.0 || >3.0.0"
      - "^0 || ^3 || ^5"

    For more information on semver, please see https://semver.org/
```